standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClientWithLocalCache.java

```diff
@@ -252,18 +264,86 @@ protected PartitionsSpecByExprResult getPartitionsSpecByExprResult(PartitionsByE
     return r;
   }
 
+  @Override
+  protected Table getTableFromLocalCache(GetTableRequest req) throws TException{
+    Table t;
+
+    if (isCacheEnabledAndInitialized() && isRequestCacheable(req, KeyType.GET_TABLE)) {
+      CacheKey cacheKey = new CacheKey(KeyType.GET_TABLE, req);
+      try {
+        t = (Table) mscLocalCache.get(cacheKey, this::load);
+
+        if (LOG.isDebugEnabled() && RECORD_STATS) {
+          LOG.debug(cacheObjName + ": " + mscLocalCache.stats().toString());
+        }
+      } catch (UncheckedCacheException e) {
+        if (e.getCause() instanceof MetaException) {
+          throw (MetaException) e.getCause();
+        } else if (e.getCause() instanceof TException) {
+          throw (TException) e.getCause();
+        } else {
+          throw new TException(e.getCause());
+        }
+      }
+    } else {
+      t =  super.getTableFromLocalCache(req);
+    }
+
+    return t;
+  }
+
+  @Override
+  protected AggrStats getAggrStatsFromLocalCache(PartitionsStatsRequest req) throws TException {
+    AggrStats r;
+
+    if (isCacheEnabledAndInitialized() && isRequestCacheable(req, KeyType.AGGR_COL_STATS)) {
+      CacheKey cacheKey = new CacheKey(KeyType.AGGR_COL_STATS, req);
+      try {
+        r = (AggrStats) mscLocalCache.get(cacheKey, this::load);
+
+        if (LOG.isDebugEnabled() && RECORD_STATS) {
+          LOG.debug(cacheObjName + ": " + mscLocalCache.stats().toString());
+        }
+      } catch (UncheckedCacheException e) {
+        if (e.getCause() instanceof MetaException) {
+          throw (MetaException) e.getCause();
+        } else if (e.getCause() instanceof TException) {
+          throw (TException) e.getCause();
+        } else {
+          throw new TException(e.getCause());
+        }
+      }
+    } else {
+      r = super.getAggrStatsFromLocalCache(req);
+    }
+
+    return r;
+  }
+
   /**
    * This method determines if the request should be cached.
    * @param request Request object
    * @return boolean
    */
-  private boolean isRequestCachable(Object request, KeyType keyType) {
+  private boolean isRequestCacheable(Object request, KeyType keyType) {
     switch (keyType) {
       //cache only requests for transactional tables, with a valid table id
       case PARTITIONS_BY_EXPR:
       case PARTITIONS_SPEC_BY_EXPR:
         PartitionsByExprRequest req = (PartitionsByExprRequest) request;
         return req.getValidWriteIdList() != null && req.getId() != -1;
+      case GET_TABLE:
+        GetTableRequest gtr = (GetTableRequest) request;
+        return gtr.getValidWriteIdList() != null;
+      case AGGR_COL_STATS:
+        PartitionsStatsRequest psr = (PartitionsStatsRequest) request;
+        Table tbl = null;
+        try {
+          tbl = getTable(psr.getDbName(), psr.getTblName());
+        } catch (TException e) {
+          return false;
+        }
+        return tbl != null && tbl.getId() != -1 && psr.getValidWriteIdList() != null;
```

Comment:
It seems `getValidWriteIdList` has a different purpose in these requests that has to do with stats retrieval.

I think it's easier to move away from relying on it at this stage. Instead, we could retrieve the `writeIdList` using `AcidUtils.TableSnapshot tableSnapshot = AcidUtils.getTableSnapshot(conf, tbl);`, and create an object in the cache that stores the `writeIdList` and the `PartitionsStatsRequest` separately.